swarmrl/agents/lymburn_model.py

```diff
@@ -0,0 +1,200 @@
+                    )
+                else:
+                    pred_force = self.pred_movement(
+                        self.t,
+                        colloid.pos,
+                        colloid.velocity,
+                        self.home_pos,
+                        self.pred_params,
+                    )
+
+                nd = np.array([pred_force[0], pred_force[1], pred_force[2]])
+                new_direction = nd / np.linalg.norm(nd)
+                actions.append(
+                    Action(force=np.linalg.norm(nd), new_direction=new_direction)
+                )
+                continue
+
+            other_colls = [c for c in colloids if c is not colloid and not c.type == 1]
+            colls_in_vision = get_colloids_in_vision(
+                colloid, other_colls, vision_radius=self.detection_radius_position_colls
+            )
+            predator = [p for p in colloids if p.type == 1]
+            # only one predator in the simulation
+            pred_in_vision = get_colloids_in_vision(
+                colloid, predator, vision_radius=self.detection_radius_position_pred
+            )
+            colls_in_vision_position = np.array([c.pos for c in colls_in_vision])
+            colls_in_vision_velocity = np.array([c.velocity for c in colls_in_vision])
+
+            pred_in_vision_position = np.array([p.pos for p in pred_in_vision])
+
+            force_a, force_r = np.array([0, 0, 0]), np.array([0, 0, 0])
+            if len(colls_in_vision) > 0:
+                force_a = np.sum(colls_in_vision_velocity - colloid.velocity, axis=0)
+
+                force_r_notnorm = np.sum(colls_in_vision_position - colloid.pos, axis=0)
+                dist_norm = np.linalg.norm(colls_in_vision_position - colloid.pos)
+                force_r = force_r_notnorm / dist_norm
+
+            force_h = self.home_pos - colloid.pos
+
+            force_p = np.array([0, 0, 0])
+            if len(pred_in_vision) > 0:
+                force_p_notnorm = np.sum(colloid.pos - pred_in_vision_position, axis=0)
+                dist_norm_pred = np.linalg.norm(colloid.pos - pred_in_vision_position)
+                force_p = force_p_notnorm / dist_norm_pred
+
+            force_f = (
+                -colloid.velocity
+                * (np.abs(colloid.velocity) - self.agent_speed)
+                / self.agent_speed
+            )
+
+            force = (
+                self.force_params["K_a"] * force_a
+                + self.force_params["K_r"] * force_r
+                + self.force_params["K_h"] * force_h
+                + self.force_params["K_p"] * force_p
+                + self.force_params["K_f"] * force_f
+            )
+
+            force_magnitude = np.linalg.norm(force)
+            force_direction = force / force_magnitude
+
+            actions.append(Action(force=force_magnitude, new_direction=force_direction))
+        return actions
+
+
+def get_colloids_in_vision(coll, other_coll, vision_radius):
```

Comment:
What are all of these functions down here for?

Reply:
It would be really cool if you could move these out of there and into the utils module still.